File: day5/typescript/p1.ts
import { readFileSync } from 'fs';


const data = readFileSync('input.txt', 'utf8');

const [depsRaw, topRaws] = data.split('\n\n');

const deps: string[] = depsRaw.split(`\n`).filter(line => line.trim() !== '');
const tops: string[] = topRaws.split(`\n`).filter(line => line.trim() !== '');
const validTops: string[] = [];


// Each supposed topological order 
for (const top of tops) {
  const orderFiltered = top.split(",");
  // J is every number of the topological order
  let viable = true;
  for (let j = 0; j < orderFiltered.length; j++) {
    // iterate over each node to find if has any Dependencies
    let stop = false;
    for (const dependence of deps) {
      const [num1, num2] = dependence.split("|");
      if (num2 === orderFiltered[j]) {
        // The number has a dependence
        let found = false;
        let foundIndex: number | null = null;

        for (let k = 0; k < orderFiltered.length; k++) {
          if (orderFiltered[k] === num1) {
            found = true
            foundIndex = k;
            break
          }
        }

        if (found && foundIndex !== null) {
          if (foundIndex < j) {
            viable = true;
          }

          if (foundIndex > j) {
            viable = false
            break;
          }
        }
      }
    }

    if (!viable) break;
  }

  if (viable) {
    validTops.push(top)
  }
}

let acc = 0;

for (const top of validTops) {
  const nums = top.split(",");
  const middle = nums[Math.floor(nums.length / 2)];
  acc += parseInt(middle)
}

console.log("ValidTops", acc)

File: day5/lua/p1.lua
local function readFile(filename)
	local file = io.open(filename, "r")
	if not file then
		error("Erro ao abrir o arquivo: " .. filename)
	end
	local data = file:read("*all")
	file:close()
	return data
end

local function split(input, delimiter)
	local result = {}
	for match in (input .. delimiter):gmatch("(.-)" .. delimiter) do
		table.insert(result, match)
	end
	return result
end

local function validateOrder(order, deps)
	for j = 1, #order do
		for _, dep in ipairs(deps) do
			local num1, num2 = dep:match("([^|]+)|([^|]+)")
			num1 = num1 and num1:match("^%s*(.-)%s*$")
			num2 = num2 and num2:match("^%s*(.-)%s*$")

			if num2 == order[j] then
				local foundIndex = nil
				for k = 1, #order do
					if order[k] == num1 then
						foundIndex = k
						break
					end
				end

				if foundIndex then
					if foundIndex > j then
						return false
					end
				end
			end
		end
	end
	return true
end

local data = readFile("input.txt")
local parts = split(data, "\n\n")
if #parts < 2 then
	error("Formato inválido do arquivo")
end

local deps = split(parts[1], "\n")
local tops = split(parts[2], "\n")
local validTops = {}

for _, top in ipairs(tops) do
	local orderFiltered = split(top, ",")
	local viable = validateOrder(orderFiltered, deps)

	if viable then
		table.insert(validTops, top)
	end
end

local acc = 0
for _, top in ipairs(validTops) do
	local nums = split(top, ",")
	local middle = nums[math.floor(#nums / 2) + 1]
	local middleValue = tonumber(middle)
	if middleValue then
		acc = acc + middleValue
	else
		print("Erro ao converter número:", middle)
	end
end

print("ValidTops:", acc)
